store/tikv/2pc.go

```diff
@@ -719,30 +745,37 @@ func (c *twoPhaseCommitter) execute(ctx context.Context) (err error) {
 	// strip check_not_exists keys that no need to commit.
 	c.stripNoNeedCommitKeys()
 
-	start = time.Now()
-	logutil.Event(ctx, "start get commit ts")
-	commitTS, err := c.store.getTimestampWithRetry(NewBackofferWithVars(ctx, tsoMaxBackoff, c.txn.vars))
+	commitTS, err := c.getCommitTS(ctx, commitDetail)
 	if err != nil {
-		logutil.Logger(ctx).Warn("2PC get commitTS failed",
-			zap.Error(err),
-			zap.Uint64("txnStartTS", c.startTS))
 		return errors.Trace(err)
 	}
-	commitDetail.GetCommitTsTime = time.Since(start)
-	logutil.Event(ctx, "finish get commit ts")
-	logutil.SetTag(ctx, "commitTs", commitTS)
 
-	// check commitTS
-	if commitTS <= c.startTS {
-		err = errors.Errorf("conn %d Invalid transaction tso with txnStartTS=%v while txnCommitTS=%v",
-			c.connID, c.startTS, commitTS)
-		logutil.BgLogger().Error("invalid transaction", zap.Error(err))
-		return errors.Trace(err)
+	tryAmend := c.isPessimistic && c.connID > 0
+	if !tryAmend {
+		_, _, err = c.checkSchemaValid(ctx, commitTS, c.txn.txnInfoSchema, false)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	} else {
+		relatedSchemaChange, memAmended, err := c.checkSchemaValid(ctx, commitTS, c.txn.txnInfoSchema, true)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if memAmended {
+			// Get new commitTS and check schema valid again.
+			newCommitTS, err := c.getCommitTS(ctx, commitDetail)
```

Comment:
Why changing the commitTS after amend?

Reply:
The prewrite operation must be done before getting `commitTS` to keep transaction consistency, the schema diff is checked using `commitTS` and amend prewrite will be done, so a new `commitTS` will be fetched and schema version will be checked again.